### PR TITLE
Reimplement HTMLHyperlinkElementUtils, Location, and History

### PR DESCRIPTION
**Pull request by domenic**
Mostly HTMLHyperlinkElementUtils for now, just so you can see how it works. Depends on https://github.com/jsdom/whatwg-url/pull/36 and still lots to do before it's mergable, including Location and getting rid of a few other uses of the old API (e.g. in History).

This should solve the performance issues since we no longer react to every `href` attribute change.

### CODE REVIEW

**Comment by rgbkrk** [comment]
Would these also make `URL.createObjectURL()` available in JSDom?

**Comment by domenic** [comment]
No

**Comment by domenic** [comment]
@Sebmaster this is ready for review, although it depends on https://github.com/jsdom/whatwg-url/pull/36

File: lib/jsdom/browser/Window.js [hunk 1]
```diff
@@ -384,6 +392,7 @@ function Window(options) {
 
   ///// INITIALIZATION
 
+  // TODO: delay the load event until everything loads.
```
Comment (Sebmaster) [inline lib/jsdom/browser/Window.js]: Does this still apply?

File: lib/jsdom/living/index.js [hunk 2]
```diff
@@ -27,6 +28,8 @@ exports.MutationEvent = require("./generated/MutationEvent").interface;
 exports.ProgressEvent = require("./generated/ProgressEvent").interface;
 exports.EventTarget = require("./generated/EventTarget").interface;
 
+exports.Location = require("./generated/Location").interface;
```
Comment (Sebmaster) [inline lib/jsdom/living/index.js]: Can we add `window.History` too?

File: lib/jsdom/browser/Window.js (hunk 1 above)
Comment (domenic) [inline lib/jsdom/browser/Window.js]: Yeah I think I misread this code; can remove.

File: lib/jsdom/living/index.js (hunk 2 above)
Comment (domenic) [inline lib/jsdom/living/index.js]: Good call

**Comment by domenic** [comment]
@Sebmaster rebased with review comments applied.

**Comment by Sebmaster** [comment]
Still missing the spec comparison, will get that done tonight too.

File: lib/jsdom/living/window/Location-impl.js [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  }
+
+  _locationObjectSetterNavigate(url) {
+    // Extra steps here to determine replacement flag are not implemented, and are not applicable to our rudimentary
+    // "navigation" implementation.
+
+    return this._locationObjectNavigate(url);
+  }
+
+  _locationObjectNavigate(url) {
```
Comment (Sebmaster) [inline lib/jsdom/living/window/Location-impl.js]: Doesn't accept the options object with replacement?

**Comment by Sebmaster** [comment]
Okay, went over the anchor and location code (except getters/setters, I assume those were mainly just copying anyways). Had a look over history and didn't notice anything, but didn't compare with spec as I assume that wouldn't make a lot of sense anyways with lots of stuff missing.

**Comment by Sebmaster** [comment]
I might be missing something, but it looks to me like the url isn't being reparsed after href has been set (if the URL is non-special). I may have to try to repro to get why it doesn't work.

**Comment by Sebmaster** [comment]
Here we go:
```js
"use strict";

const jsdom = require(".");

const document = jsdom.jsdom("<a href='mycustomprotocol:abc'>Hello</a>");

const anchor = document.getElementsByTagName("a")[0];

console.log(anchor.protocol);
anchor.href = 'otherprot:test.js';

console.log(anchor.protocol);
```

Doesn't do what you'd expect. Works with the URL constructor however.

File: lib/jsdom/living/window/Location-impl.js (hunk 3 above)
Comment (domenic) [inline lib/jsdom/living/window/Location-impl.js]: It does, but see the comment. We don't support replacement.

Comment (domenic) [inline lib/jsdom/living/window/Location-impl.js]: I will add a commented out argument and keep the call site. What do you think?

**Comment by domenic** [comment]
Nice catch; will add a test. I think it is covered by the clause

> When elements implementing the HTMLHyperlinkElementUtils mixin are created, and whenever those elements have their href content attribute set, changed, or removed, the user agent must set the url.

Which we don't implement, since the following note says 

> This is only observable for blob: URLs as parsing them involves the structured clone algorithm.

I guess that note is not completely true. We still don't want to implement it though due to the perf consequences during parsing.

I will just add an explicit call in the setter.

File: lib/jsdom/living/window/Location-impl.js (hunk 3 above)
Comment (Sebmaster) [inline lib/jsdom/living/window/Location-impl.js]: :+1:

**Comment by domenic** [comment]
OK, fixed in a slightly different way since it is indeed a spec bug, in a separate commit on top of the others.

**Pull request merged**